staging/src/k8s.io/apiserver/pkg/util/flowcontrol/rule.go

```diff
@@ -0,0 +1,129 @@
+	"k8s.io/apiserver/pkg/authentication/serviceaccount"
+)
+
+func matchesFlowSchema(digest RequestDigest, flowSchema *rmtypesv1alpha1.FlowSchema) bool {
+	for _, policyRule := range flowSchema.Spec.Rules {
+		subjectMatches := false
+		for _, subject := range policyRule.Subjects {
+			if matchesSubject(digest, subject) {
+				subjectMatches = true
+			}
+		}
+		if !subjectMatches {
+			continue
+		}
+
+		if policyRuleMatches := matchesPolicyRule(digest, policyRule.Rule); policyRuleMatches {
+			return true
+		}
+	}
+	return false
+}
+
+func matchesSubject(digest RequestDigest, subject rmtypesv1alpha1.Subject) bool {
+	user := digest.User
+	switch subject.Kind {
+	case rmtypesv1alpha1.UserKind:
+		return user.GetName() == subject.Name
+	case rmtypesv1alpha1.GroupKind:
+		matches := false
+		foreach(user.GetGroups(), func(group string) { matches = matches || group == subject.APIGroup })
+		return matches
+	case rmtypesv1alpha1.ServiceAccountKind:
+		return serviceaccount.MatchesUsername(subject.Namespace, subject.Name, user.GetName())
+	default:
+		return false
+	}
+}
+
+func matchesPolicyRule(digest RequestDigest, policyRule rmtypesv1alpha1.PolicyRule) bool {
+	if !matchPolicyRuleVerb(policyRule.Verbs, digest.RequestInfo.Verb) {
+		return false
+	}
+	if !digest.RequestInfo.IsResourceRequest {
+		return matchPolicyRuleNonResourceURL(policyRule.NonResourceURLs, digest.RequestInfo.Path)
+	}
+	if !matchPolicyRuleResource(policyRule.Resources, digest.RequestInfo.Resource) {
+		return false
+	}
+	if !matchPolicyRuleAPIGroup(policyRule.APIGroups, digest.RequestInfo.APIGroup) {
+		return false
+	}
+	return true
+}
+
+func matchPolicyRuleVerb(policyRuleVerbs []string, requestVerb string) bool {
+	if containsWildcard(policyRuleVerbs) {
+		return true
+	}
+	matched := false
+	foreach(policyRuleVerbs, func(verb string) { matched = matched || verb == requestVerb })
+	return matched
+}
+
+func matchPolicyRuleNonResourceURL(policyRuleRequestURLs []string, requestPath string) bool {
+	if containsWildcard(policyRuleRequestURLs) {
+		return true
+	}
+	matched := false
+	foreach(policyRuleRequestURLs, func(path string) {
+		matched = matched || path == requestPath
+	})
+	return matched
+}
+
+func matchPolicyRuleAPIGroup(policyRuleAPIGroups []string, requestAPIGroup string) bool {
+	if containsWildcard(policyRuleAPIGroups) {
+		return true
+	}
+	matched := false
+	foreach(policyRuleAPIGroups, func(group string) {
+		matched = matched || group == requestAPIGroup
+	})
+	return matched
+}
+
+func matchPolicyRuleResource(policyRuleRequestResources []string, requestResource string) bool {
+	if containsWildcard(policyRuleRequestResources) {
+		return true
+	}
+	matched := false
+	foreach(policyRuleRequestResources, func(resource string) {
+		matched = matched || resource == requestResource
+	})
+	return matched
+}
+
+func containsWildcard(ruleItems []string) bool {
+	contains := false
+	foreach(ruleItems, func(item string) {
```

Comment:
Do we need a loop here?  I thought our validation promises the wildcard is the first and only element, if there is a wildcard.

Reply:
refactored